tests/test_asttokens.py

```diff
@@ -34,6 +36,38 @@ def test_tokenizing(self):
     self.assertEqual(atok.tokens[5].endpos, 25)
 
 
+  def test_given_existing_tokens(self):
```

Comment:
This is almost identical to the previous test, which makes sense. Please deduplicate the code.

It would also be nice to have a test of what happens when you pass tokens that aren't identical to what's produced normally.

Reply:
> It would also be nice to have a test of what happens when you pass tokens that aren't identical to what's produced normally.

I'm not sure I follow; do you have an example which would clarify this?
Do you mean for example passing incorrect tokens which don't match the source? Or something else?

Reply:
Incorrect tokens would do the trick, I can't think of any other way. I partly just want something that shows that it really used the passed in tokens rather than generating its own.

Reply:
https://github.com/gristlabs/asttokens/pull/95/commits/363c1323ba6183dd9b7060f63edb22986bf59575 deduplicates, https://github.com/gristlabs/asttokens/pull/95/commits/fa69685b81841f3b249571352b70bc8241ad9b0f validates the tokens are used.